ui/src/treegrid.tsx

```diff
@@ -0,0 +1,247 @@
+import ErrorIcon from "@mui/icons-material/Error";
+import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
+import KeyboardArrowRightIcon from "@mui/icons-material/KeyboardArrowRight";
+import CircularProgress from "@mui/material/CircularProgress";
+import IconButton from "@mui/material/IconButton";
+import Typography from "@mui/material/Typography";
+import { ReactNode, useCallback } from "react";
+import { useImmer } from "use-immer";
+
+export type TreeGridId = string | number;
+
+export type TreeGridRowData = {
+  [key: string]: string | number | JSX.Element;
```

Comment:
Another approach you could consider is to move an optional rendering function into the Column definition. This would require less data prep by the client to pass in the new values for `rows`, as it doesn't need to worry about display concerns.